# SOEN363 Project

## Drop the DB Indecies **(Only for Debugging Purposes)**

In [ ]:
DROP INDEX actors_index;

## Drop the DB Views **(Only for Debugging Purposes)**

In [ ]:
DROP VIEW high_ratings;

DROP VIEW low_ratings;

## Drop the DB tables **(Only for Debugging Purposes)**

In [ ]:
DROP TABLE actors;

DROP TABLE genres;

DROP TABLE tags;

DROP TABLE tag_names;

DROP TABLE movies;

# Assignment Questions
## 2a

In [ ]:
CREATE TABLE movies (
    mid             INTEGER UNIQUE, 
    title           VARCHAR, 
    year            INTEGER, 
    rating          REAL CHECK (rating BETWEEN 0.0 AND 5.0), 
    num_ratings     INTEGER,
    PRIMARY KEY     (mid, title, year)
);

CREATE TABLE actors (
    mid             INTEGER REFERENCES movies(mid), 
    name            VARCHAR, 
    cast_position   INTEGER,
    PRIMARY KEY     (mid, name)
);

CREATE TABLE genres (
    mid             INTEGER REFERENCES movies(mid), 
    genre           VARCHAR,
    PRIMARY KEY     (mid, genre)
);

CREATE TABLE tag_names (
    tid             INTEGER, 
    tag             VARCHAR,
    PRIMARY KEY     (tid)
);

CREATE TABLE tags (
    mid             INTEGER REFERENCES movies(mid), 
    tid             INTEGER,
    PRIMARY KEY     (mid, tid),
    FOREIGN KEY     (tid) REFERENCES tag_names(tid)
);

COPY movies FROM '/srv/soen363/phase1/movies.dat';

COPY actors FROM '/srv/soen363/phase1/actors.dat';

COPY genres FROM '/srv/soen363/phase1/genres.dat';

COPY tag_names FROM '/srv/soen363/phase1/tag_names.dat';

COPY tags FROM '/srv/soen363/phase1/tags.dat';

## 3a

In [ ]:
-- Get the movie titles starring "Daniel Craig", sorted in an ascending alphabetical order
SELECT DISTINCT M.title AS "Movie Titles"
FROM actors A, movies M
WHERE
    M.mid = A.mid AND
    A.name = 'Daniel Craig'
ORDER BY M.title ASC;

(19 row(s) affected)

Total execution time: 00:00:00.183

Movie Titles A Kid in King Arthur's Court Archangel Casino Royale Elizabeth Enduring Love Infamous Lara Croft: Tomb Raider Layer Cake Munich Quantum of Solace Renaissance Road to Perdition Sorstalanság Sylvia The Golden Compass The Invasion The Jacket The Mother The Power of One

## 3b

In [ ]:
-- Get the cast of the movie titled "The Dark Knight", sorted in ascending alphabetical order
SELECT DISTINCT A.name AS "Cast"
FROM actors A, movies M
WHERE
    M.mid = A.mid AND
    M.title = 'The Dark Knight'
ORDER BY A.name ASC;

(91 row(s) affected)

Total execution time: 00:00:00.013

Cast Aaron Eckhart Adam Kalesperis Aidan Feore Andrew Bicknell Andy Luther Anthony Michael Hall Ariyon Bakare Beatrice Rosen Bill Smille Brandon Lambdin Bronson Webb Chin Han Christian Bale Chucky Venn Cillian Murphy Colin McFarlane Craig Heaney Crhis Perschler Dale RIvera Daryl Satcher David Dastmalchian Doug Ballard Edison Chen Eric Roberts Erik Hellman Gary Oldman Gertrude Kyles Grahame Edwards Greg Beam Hannah Gunn Heath Ledger Helene Wilson Ian Pirie James Farruggio James Fierro James Scales Jennifer Knox Jonathan Ryland Joseph Luis Caballero Joshua Harto Joshua Rollins Keith Kupferer K. Todd Freeman Lanny Lutz Lateef Lovejoy Lorna Gayle Maggie Gyllenhaal Matthew Leitch Matthew O'Neill Matt Rippy Matt Shallenberger Melinda McGraw Michael Andrew Gorman Michael Caine Michael Corey Foster Michael Jai White Michael Stoyanov Michael Vieau Monique Curnen Morgan Freeman Nancy Crane Nathan Gamble Nestor Carbonell Nigel Carrington Nydia Rodriguez Terracina Olumiji Olawumi Patrick Clear Paul Birchard Peter DeFaria Philip Bulcock Richard Dillane Ritchie Coster Roger Monk Ronan Summers Ron Dean Sam Derence Sarah Jayne Dunn Sophia Hinshelwood Thomas Gaitsch Thomas McElroy Tommy Campbell Tommy 'Tiny' Lister Tristan Tait Vincent Riotta Vincenzo Nicoli Wai Wong Walter Lewis William Armstrong William Fichtner Will Zahrn Winston Ellis

## 3c

In [ ]:
-- Get the distinct genres in the database, alongside their number of occurrences (only for 1000+ occurrences), sorted ascending numerically in occurrences
SELECT DISTINCT G.genre, COUNT(M.*) AS number_of_movies_in_genre
FROM genres G, movies M
WHERE G.mid = M.mid
GROUP BY G.genre
HAVING COUNT(M.*) > 1000
ORDER BY COUNT(M.*) ASC;

(7 row(s) affected)

Total execution time: 00:00:00.037

genre number_of_movies_in_genre Adventure 1003 Crime 1086 Action 1445 Romance 1644 Thriller 1664 Comedy 3566 Drama 5076

## 3d

In [ ]:
-- For each year, print the movie title, year, and rating, sorted in the ascending order of year and the descending order of movie rating
SELECT M.title, M.year, M.rating
FROM movies M
WHERE M.num_ratings > 0999
ORDER BY M.year ASC, M.rating DESC;

(7346 row(s) affected)

Displaying Top 5000 rows.

Total execution time: 00:00:00.085

title year rating The Birth of a Nation 1915 3.3 Intolerance: Love's Struggle Throughout the Ages 1916 3.8 Broken Blossoms or The Yellow Man and the Girl 1919 3.7 Das Cabinet des Dr. Caligari. 1920 4.1 Dr. Mabuse, der Spieler - Ein Bild der Zeit 1922 4.1 Dr. Mabuse, der Spieler - Ein Bild der Zeit 1922 4.1 Nosferatu, eine Symphonie des Grauens 1922 3.9 Häxan 1922 3.8 Nanook of the North 1922 3.7 Der letzte Mann 1924 4.1 The Thief of Bagdad 1924 3.6 The Thief of Bagdad 1924 3.6 The Thief of Bagdad 1924 3.6 The Gold Rush 1925 4.1 Bronenosets Potyomkin 1925 4.0 The Phantom of the Opera 1925 3.7 The General 1926 4.4 The General 1926 4.4 The General 1926 4.4 Battling Butler 1926 3.7 Sunrise: A Song of Two Humans 1927 4.4 Metropolis 1927 4.1 Metropolis 1927 4.1 Berlin: Die Sinfonie der Grosstadt 1927 3.9 It 1927 3.7 La passion de Jeanne d'Arc 1928 4.4 The Circus 1928 4.1 Oktyabr 1928 4.0 Un chien andalou 1929 4.1 Die Büchse der Pandora 1929 4.1 Die Büchse der Pandora 1929 4.1 Chelovek s kino-apparatom 1929 4.1 The Cocoanuts 1929 3.6 Blackmail 1929 3.4 The Manxman 1929 3.1 The Broadway Melody 1929 2.9 Der blaue Engel 1930 4.0 L'âge d'or 1930 4.0 Le sang d'un poète 1930 3.9 All Quiet on the Western Front 1930 3.8 All Quiet on the Western Front 1930 3.8 The Bat Whispers 1930 3.4 City Lights 1931 4.3 City Lights 1931 4.3 M 1931 4.3 The Public Enemy 1931 3.9 Tabu: A Story of the South Seas 1931 3.8 À nous la liberté 1931 3.7 Dracula 1931 3.6 Little Caesar 1931 3.6 Frankenstein 1931 3.6 Dracula 1931 3.6 Dracula 1931 3.6 Frankenstein 1931 3.6 The Skin Game 1931 2.7 Trouble in Paradise 1932 4.2 Freaks 1932 4.0 Horse Feathers 1932 3.9 Grand Hotel 1932 3.8 Duck Soup 1933 4.1 The Invisible Man 1933 3.7 42nd Street 1933 3.5 Cavalcade 1933 3.0 L'atalante 1934 4.3 The Thin Man 1934 4.2 It Happened One Night 1934 4.1 Manhattan Melodrama 1934 3.7 Twentieth Century 1934 3.7 Babes in Toyland 1934 3.6 Babes in Toyland 1934 3.6 Man of Aran 1934 3.6 A Night at the Opera 1935 4.1 Top Hat 1935 4.0 The 39 Steps 1935 3.8 Mutiny on the Bounty 1935 3.7 Mutiny on the Bounty 1935 3.7 Bride of Frankenstein 1935 3.7 Alice Adams 1935 3.6 Barbary Coast 1935 3.1 Modern Times 1936 4.2 Modern Times 1936 4.2 My Man Godfrey 1936 4.1 My Man Godfrey 1936 4.1 Dodsworth 1936 4.0 Camille 1936 3.9 Sabotage 1936 3.5 The Great Ziegfeld 1936 3.4 Things to Come 1936 3.3 Come and Get It 1936 3.1 Tell Your Children 1936 3.0 La grande illusion 1937 4.3 The Awful Truth 1937 4.0 Pépé le Moko 1937 3.8 Drôle de drame ou L'étrange aventure du Docteur Molyneux 1937 3.8 The Good Earth 1937 3.7 The Life of Emile Zola 1937 3.6 A Damsel in Distress 1937 3.5 Young and Innocent 1937 3.4 Le quai des brumes 1938 4.1 La bête humaine 1938 4.0 Aleksandr Nevskiy 1938 4.0 The Lady Vanishes 1938 4.0 Bringing Up Baby 1938 4.0 You Can't Take It with You 1938 4.0 Pygmalion 1938 3.7 The Adventures of Robin Hood 1938 3.7 The Adventures of Robin Hood 1938 3.7 La règle du jeu 1939 4.3 Ninotchka 1939 4.1 Young Mr. Lincoln 1939 3.9 Dark Victory 1939 3.9 Wuthering Heights 1939 3.9 Gone with the Wind 1939 3.9 Wuthering Heights 1939 3.9 Wuthering Heights 1939 3.9 Only Angels Have Wings 1939 3.8 Gunga Din 1939 3.7 The Wizard of Oz 1939 3.6 Mr. Smith Goes to Washington 1939 3.6 Jamaica Inn 1939 3.0 The Great Dictator 1940 4.2 The Philadelphia Story 1940 4.1 The Shop Around the Corner 1940 4.1 Waterloo Bridge 1940 4.0 Christmas in July 1940 4.0 My Favorite Wife 1940 3.9 The Grapes of Wrath 1940 3.8 Foreign Correspondent 1940 3.7 Fantasia 1940 3.5 Pinocchio 1940 3.3 Pinocchio 1940 3.3 Sullivan's Travels 1941 4.2 Sullivan's Travels 1941 4.2 The Lady Eve 1941 4.1 Ball of Fire 1941 4.1 Citizen Kane 1941 4.1 The Maltese Falcon 1941 4.0 Sergeant York 1941 4.0 The Maltese Falcon 1941 4.0 How Green Was My Valley 1941 3.9 The Wolf Man 1941 3.7 Penny Serenade 1941 3.7 Johnny Eager 1941 3.7 Dumbo 1941 3.3 To Be or Not to Be 1942 4.2 To Be or Not to Be 1942 4.2 The Magnificent Ambersons 1942 4.1 The Major and the Minor 1942 4.0 

## 3e

In [ ]:
-- Get the titles of movies whose tags include at least one occurrence of the word "bad" and another of "good"
SELECT M.title
FROM movies M, tags T, tag_names TN
WHERE 
    M.mid = T.mid AND T.tid = TN.tid AND
    TN.tag IN (
        SELECT TN.tag 
        FROM tag_names TN
        WHERE TN.tag LIKE 'good%'
        UNION
        SELECT TN.tag 
        FROM tag_names TN
        WHERE TN.tag LIKE 'bad%'
    );

(206 row(s) affected)

Total execution time: 00:00:00.048

title Chung Hing sam lam Down Periscope Bad Boys Die Hard: With a Vengeance Nine Months Clerks. Houseguest Miracle on 34th Street Pulp Fiction Pulp Fiction While You Were Sleeping Forrest Gump Forrest Gump The Piano Sleepless in Seattle Sliver Super Mario Bros. The Nightmare Before Christmas Ghost Aladdin Snow White and the Seven Dwarfs Pretty Woman Heavy Metal Mulholland Falls Independence Day Matilda Breakfast at Tiffany's My Fair Lady Gigi The Sound of Music The Sound of Music E.T.: The Extra-Terrestrial Microcosmos: Le peuple de l'herbe Monty Python and the Holy Grail The Princess Bride Raiders of the Lost Ark Lawrence of Arabia Lawrence of Arabia Amadeus The Shining The Manchurian Candidate Ben-Hur Pink Floyd The Wall When Harry Met Sally... Alien³ Bride of Frankenstein Cape Fear The Omen Star Trek VI: The Undiscovered Country Grease Grosse Pointe Blank Speed 2: Cruise Control Men in Black Good Will Hunting Scream 2 Titanic Wag the Dog Bulworth A Perfect Murder Labyrinth Back to the Future Part II The Last Temptation of Christ The Black Cauldron Doctor Zhivago Indiana Jones and the Temple of Doom Ever After Beetle Juice Edward Scissorhands What Dreams May Come La vita è bella First Blood You've Got Mail Patch Adams The King and I Never Been Kissed SLC Punk! eXistenZ Dracula War of the Worlds War of the Worlds Notting Hill Notting Hill Lola rennt Runaway Bride The Sixth Sense Excalibur Risky Business Per un pugno di dollari Home Alone 2: Lost in New York Time Bandits Bringing Out the Dead House on Haunted Hill The Shop Around the Corner Awakenings Stuart Little 2 Titus The Cutting Edge Wonder Boys The Ninth Gate Teenage Mutant Ninja Turtles The Skulls Small Time Crooks Mad Max 2 Mad Max 2 Prince of the City The Adventures of Rocky & Bullwinkle F/X2 Almost Famous Bedazzled Dude, Where's My Car? Annie Memento Memento Bridget Jones's Diary Unlawful Entry Faust: Love of the Damned Legally Blonde Tango & Cash Serendipity Mulholland Dr. Harry Potter and the Sorcerer's Stone Le fabuleux destin d'Amélie Poulain Kate & Leopold The Lord of the Rings: The Fellowship of the Ring The Flight of Dragons Impostor The Count of Monte Cristo Vampire Hunter D: Bloodlust Rashômon Rashômon Minority Report Sen to Chihiro no kamikakushi Sweet Home Alabama The Ring Frida The Pianist How to Lose a Guy in 10 Days Daredevil Daredevil Bend It Like Beckham Cowboy Bebop: Tengoku no tobira Mies vailla menneisyyttä Le peuple migrateur Confidence A Chorus Line Evan Almighty Bad Boys II Commando Underworld Runaway Jury The Matrix Revolutions The Public Enemy Red River Mona Lisa Smile Teacher's Pet Hangman's Curse After the Thin Man Shrek 2 Cellular The Forgotten Ray Bridget Jones: The Edge of Reason Ocean's Twelve Prelude to War The Sleeping Dictionary The Sleeping Dictionary The Animatrix A Scanner Darkly Walk on Water Crash Star Wars: Episode III - Revenge of the Sith Batman Begins Bewitched Die fetten Jahre sind vorbei Fantastic Four Must Love Dogs Serenity Just Like Heaven Zathura: A Space Adventure Movie Crazy The Libertine The Chronicles of Narnia: The Lion, the Witch and the Wardrobe Bandidas V for Vendetta V for Vendetta Mission: Impossible III Clerks II Lady in the Water Superman Returns Miami Vice The Illusionist The Queen El laberinto del fauno The Departed Notes on a Scandal Miss Potter Zodiac Shizukanaru kettô The Condemned Shrek the Third Once The Last Legion Michael Clayton 30 Days of Night I Am Legend Kung Fu Panda

## 3f

In [ ]:
-- i. Get the information for the movies with the highest number of ratings, sorted ascending by movie ID
SELECT M.mid, M.title, M.year, M.rating, M.num_ratings
FROM movies M
WHERE M.num_ratings IN (
    SELECT MAX(M.num_ratings)
    FROM movies M
)
ORDER BY M.mid;

(2 row(s) affected)

Total execution time: 00:00:00.011

mid title year rating num_ratings 4201 Pirates of the Caribbean: At World's End 2007 3.8 1768593 53125 Pirates of the Caribbean: At World's End 2007 3.8 1768593

In [ ]:
-- ii. Get the information for the movies with the highest rating, sorted ascending by movie ID
SELECT M.mid, M.title, M.year, M.rating, M.num_ratings
FROM movies M
WHERE M.rating IN (
    SELECT MAX(M.rating)
    FROM movies M
)
ORDER BY M.mid;

(1 row(s) affected)

Total execution time: 00:00:00.013

mid title year rating num_ratings 4311 1732 Høtten 1998 5.0 5

In [ ]:
-- iii. Get the information for the movies with BOTH the highest rating AND the highest number of ratings (if it exists)

Total execution time: 00:00:00

From the data returned from the last two queries, we know that this isn't the case; the movie with the highest rating isn't among the movies with the highest number of ratings.

Highest number of ratings: mid 4201/53125, title Pirates of the Caribbean: At World's End

Highest rating: mid 4311, title 1732 Høtten

In [2]:
-- iv. Get the information for the movies with the lowest rating, sorted by ascending movie ID
SELECT M.mid, M.title, M.year, M.rating, M.num_ratings
FROM movies M
WHERE M.rating IN (
    SELECT MIN(M.rating)
    FROM movies M
    WHERE M.num_ratings > 0
)
ORDER BY M.mid;

(1 row(s) affected)

Total execution time: 00:00:00.019

mid title year rating num_ratings 4230 Too Much Sleep 1997 1.5 3

In [ ]:
-- v. Get the information for the movies with BOTH the lowest AND the highest number of ratings (if it exists)

From the data returned from the last query, we know that the lowest rating (for a movie which has been rated more than 0 times) is 1.5
This means that it is not the case for the movie with the highest number of ratings (currently 1'768'593 ratings at 3.8) to be the movie with the lowest rating (currently 3 ratings at 1.5).

In [ ]:
-- vi. Present your findings on the hypothesis: are popular movies (the most ratings) among the best/worst movies (lowest or highest rating score)?

According to the results from the previous queries, it can be safely said that the hypothesis is false for this database.

## 3g

In [25]:
-- Get the year, title and rating of the lowest-rated and highest-rated movies for each year, between 2005 and 2011 inclusively. If tied, use ascending title

SELECT M2.year, M2.title, M2.rating
FROM movies M2
INNER JOIN (
    SELECT M1.year, MAX(M1.rating) as highest_rating
    FROM movies M1
    WHERE M1.year BETWEEN 2005 AND 2011 AND M1.num_ratings > 0
    GROUP BY M1.year
    ORDER BY M1.year ASC
) highest ON (highest.year = M2.year AND highest.highest_rating = M2.rating)

UNION

SELECT M4.year, M4.title, M4.rating
FROM movies M4
INNER JOIN (
    SELECT M3.year, MIN(M3.rating) as lowest_rating
    FROM movies M3
    WHERE M3.year BETWEEN 2005 AND 2011 AND M3.num_ratings > 0
    GROUP BY M3.year
    ORDER BY M3.year ASC
) lowest ON (lowest.year = M4.year AND lowest.lowest_rating = M4.rating)

ORDER BY year ASC, rating ASC, title ASC

(17 row(s) affected)

Total execution time: 00:00:00.012

year title rating 2005 Alone in the Dark 2.1 2005 Son of the Mask 2.1 2005 No Direction Home: Bob Dylan 4.3 2006 Basic Instinct 2 2.5 2006 Bug 2.5 2006 Doogal 2.5 2006 Das Leben der Anderen 4.4 2007 D-War 2.3 2007 Byôsoku 5 senchimêtoru 4.3 2007 No End in Sight 4.3 2007 Pete Seeger: The Power of Song 4.3 2007 War Dance 4.3 2008 Asylum 2.2 2008 The Dark Knight 4.5 2009 Miss March 2.7 2009 The Last House on the Left 2.7 2009 Star Trek 4.1

Only movies with more than 0 ratings were considered in this query. As no ratings were given to movies in 2010 and 2011, no movies from those years appear in the results. Removing the constraint will make the results appear, albeit with their ratings being 0.0.

## 3h

In [ ]:
-- High Ratings view to get actors that have participated in a movie with a rating >= 4
CREATE OR REPLACE VIEW high_ratings AS
    SELECT DISTINCT A.name 
    FROM actors A, movies M
    WHERE A.mid = m.mid AND m.rating >= 4; 

-- Low Ratings view to get actors that have participated in a movie with a rating < 4
CREATE OR REPLACE VIEW low_ratings AS
    SELECT DISTINCT A.name 
    FROM actors A, movies M
    WHERE A.mid = m.mid AND m.rating < 4;

-- Get the actors count coming from High Ratings 
SELECT COUNT(*) as result_count
FROM high_ratings;

-- Get the actors count coming from Low Ratings
SELECT COUNT(*) as result_count
FROM low_ratings;

In [ ]:
-- Get the number of No Flop actors in the database
SELECT COUNT(*)
FROM high_ratings
WHERE name NOT IN (
    SELECT name
    FROM low_ratings
);

In [ ]:
-- Get the name and movie count of the Top 10 No Flop actors
SELECT R.name, COUNT(A.*) as movie_count
FROM high_ratings R, actors A
WHERE 
    R.name NOT IN (
        SELECT name
        FROM low_ratings
    ) AND 
    A.name = R.NAME
GROUP BY R.name
ORDER BY COUNT(A.*) DESC
LIMIT 10;

## 3i

In [ ]:
SELECT 
    A.name, 
    MIN(M.year) as start, 
    MAX(M.year) as end, 
    MAX(M.year) - MIN(M.year) as longevity
FROM actors A, movies M
WHERE A.mid = M.mid
GROUP BY A.name 
ORDER BY MAX(M.year) - MIN(M.year) DESC
LIMIT 1;

## 3j

In [4]:
-- Returns the distinct names of actors who played in at least one movie with Annette Nicole.
CREATE VIEW co_actors AS
SELECT DISTINCT name
FROM actors 
WHERE mid IN (SELECT mid FROM actors
            WHERE name = 'Annette Nicole');

-- Print the number of rows in this view. 
SELECT COUNT(name)
FROM co_actors;

Commands completed successfully

(1 row(s) affected)

Total execution time: 00:00:00.073

count 179

In [5]:
-- Returns all possible combinations of co_actors and the movie ids in which Annette Nicole played.
CREATE VIEW all_combinations AS
SELECT DISTINCT co_actors.name, actors.mid
FROM co_actors, actors
WHERE actors.mid IN (SELECT mid FROM actors
            WHERE name = 'Annette Nicole')
ORDER BY co_actors.name;

-- Print the number of rows in this view. 
SELECT COUNT(name)
FROM all_combinations;

Commands completed successfully

(1 row(s) affected)

Total execution time: 00:00:00.186

count 537

In [6]:
-- Remove all legitimate pairs from all_combinations.
CREATE VIEW non_existent AS
SELECT * FROM all_combinations
WHERE NOT EXISTS (SELECT * 
FROM actors
WHERE all_combinations.name = actors.name AND all_combinations.mid = actors.mid);

-- Print the number of rows in this view.
SELECT COUNT(name)
FROM non_existent;

Commands completed successfully

(1 row(s) affected)

Total execution time: 00:00:00.312

count 239

In [3]:
-- Eliminate the distinct actors from co_actors that appear in the view non_existent.
-- Print the names of all co_actors except Annette Nicole.

SELECT * FROM co_actors
WHERE NOT EXISTS (SELECT * 
FROM non_existent
WHERE non_existent.name = co_actors.name) AND NOT co_actors.name = 'Annette Nicole';

(2 row(s) affected)

Total execution time: 00:00:00.242

name Christian Perry Kristen Connolly

## 3k

## 3l

## 3m

In [ ]:
-- Detecting duplicates
SELECT title, year, rating, num_ratings, COUNT(*) occurrences
FROM movies
GROUP BY title, year, rating, num_ratings
HAVING COUNT(*) > 1;

(681 row(s) affected)

Total execution time: 00:00:00.291

title year rating num_ratings occurrences Singin' in the Rain 1952 3.9 28033 3 The Last American Virgin 1982 3.3 3330 2 The Best Man 2005 0.0 0 2 Chocolat 2000 3.6 62492 2 The Bridges of Madison County 1995 3.6 11573 2 Mighty Joe Young 1998 2.8 10755 2 Bill & Ted's Bogus Journey 1991 3.0 17201 2 The Devil-Doll 1936 0.0 0 2 Gegen die Wand 2004 0.0 0 2 Stuart Little 2 2002 2.7 10893 2 Believers 2007 0.0 0 2 In Cold Blood 1967 3.8 2614 2 Joan of Arc 1999 3.1 8110 2 Dawn of the Dead 2004 3.4 49706 3 Lola 1981 3.9 529 2 Sinbad and the Eye of the Tiger 1977 0.0 0 2 Modern Times 1936 4.2 9563 2 Lemony Snicket's A Series of Unfortunate Events 2004 3.2 60437 2 Ae Fond Kiss... 2004 3.7 1269 2 Love and a Bullet 2002 0.0 0 2 The Chronicles of Narnia: Prince Caspian 2008 3.8 154104 2 V for Vendetta 2006 3.9 226041 2 La boum 1980 NULL NULL 2 La double vie de Véronique 1991 4.3 5676 2 The Shaggy Dog 2006 2.9 41890 2 Some Girls 1988 0.0 0 2 CrissCross 1992 0.0 0 2 Alien³ 1992 3.1 30526 2 Candy 2006 3.8 9542 2 Gladiator 2000 3.8 248672 2 The Adventures of Huck Finn 1993 2.9 2408 2 La decima vittima 1965 0.0 0 2 The Texas Chainsaw Massacre 2003 3.2 36030 2 O Brother, Where Art Thou? 2000 3.7 35122 2 The Business 2005 0.0 0 2 The In-Laws 2003 2.9 2546 2 Ben-Hur 1959 3.7 25817 2 Appaloosa 2008 3.2 19629 2 The Little Shop of Horrors 1960 0.0 0 2 The Postman Always Rings Twice 1981 3.3 2610 2 Darkness Falls 2003 2.8 11135 2 Contact 1997 3.1 19572 2 The Front Page 1974 3.6 905 2 Runaway Jury 2003 3.4 11578 2 Jane Eyre 2006 NULL NULL 3 Whale Rider 2002 3.8 9564 2 The Boys of Baraka 2005 4.0 627 2 Man on Fire 2004 3.9 67139 2 My Darling Clementine 1946 3.8 1741 2 Loser 2000 2.6 5745 2 Bride of Chucky 1998 2.6 20874 2 Hard Target 1993 0.0 0 2 Faust: Love of the Damned 2000 0.0 0 2 Miami Vice 2006 2.8 55637 2 High School Musical 3: Senior Year 2008 3.6 80210 2 Night of the Living Dead 1968 3.7 26113 2 Notes on a Scandal 2006 3.7 22200 2 Twisted 2004 2.8 2961 2 Harold & Kumar Go to White Castle 2004 3.5 61100 2 Phenomenon 1996 3.0 9166 2 12 Angry Men 1957 4.1 22784 2 Sahara 2005 3.0 24489 2 Bangkok Dangerous 2008 2.8 23793 2 Mutiny on the Bounty 1935 3.7 1846 2 Twelve O'Clock High 1949 0.0 0 2 Boogeyman 2005 2.5 13866 2 Being Julia 2004 3.5 1947 2 House of Wax 2005 2.8 39495 2 Punch-Drunk Love 2002 3.2 24379 2 The Mother 2003 3.4 862 2 Shichinin no samurai 1954 4.5 29509 2 Village of the Damned 1995 2.7 4194 2 Paradise 1982 0.0 0 2 The Bourne Identity 2002 3.9 125823 2 The Three Musketeers 1993 3.2 13783 3 Planet Terror 2007 0.0 0 2 Never Say Never Again 1983 3.1 7414 2 Boy Meets Girl 1984 0.0 0 2 August Rush 2007 4.0 105941 4 The Razor's Edge 1984 3.7 879 2 This Is Spinal Tap 1984 3.9 23608 3 How Stella Got Her Groove Back 1998 3.1 4439 2 DOA: Dead or Alive 2006 2.7 71044 4 Il mio viaggio in Italia 1999 3.9 769 2 Kansas City 1996 3.0 608 2 Auto Focus 2002 3.2 1283 2 Punisher: War Zone 2008 3.0 29393 4 Abandon 2002 2.7 2670 2 Q 1982 0.0 0 2 The Thomas Crown Affair 1999 3.3 15699 2 Communion 1976 0.0 0 2 The Island 2005 3.3 41199 2 King Arthur 2004 3.2 57867 2 Dangerous Liaisons 1988 3.7 11359 2 Spirit: Stallion of the Cimarron 2002 3.3 8302 2 Into the West 1992 3.6 744 2 Sunshine 2007 0.0 0 2 Interview 2007 3.2 14637 2 Man on the Moon 1999 3.3 15027 2 The Fall 2006 3.8 23397 2 A Midsummer Night's Dream 1999 0.0 0 2 Peter Pan 2003 3.3 47413 2 Charlie and the Chocolate Factory 2005 0.0 0 3 Othello 1995 3.3 2018 2 Flight of the Phoenix 2004 3.0 7566 2 That Darn Cat 1997 2.5 2029 2 The School of Rock 2003 3.2 183700 2 "Grey's Anatomy" 2005 NULL NULL 2 "Earth: Final Conflict" 1997 NULL NULL 2 The Neverending Story III 1994 0.0 0 2 The Fourth Protocol 1987 3.4 544 2 La môme 2007 3.8 33 3 The Grass Is Greener 1960 0.0 0 2 Enchanted 2007 3.8 188655 3 Spy Kids 2: Island of Lost Dreams 2002 2.6 12329 2 Dead Ringers 1988 3.6 4690 2 A Star Is Born 1937 0.0 0 2 Sleuth 2007 3.2 9724 2 Albino Alligator 1996 3.0 1198 2 Huo Yuan Jia 2006 0.0 0 2 Funny Games U.S. 2007 3.2

In [ ]:
-- Creating a view that contains no duplicates
CREATE VIEW no_duplicates AS 
SELECT DISTINCT title, year, rating, num_ratings
FROM movies
GROUP BY title, year, rating, num_ratings;

## 4a

In [ ]:
CREATE INDEX actors_index ON actors(name);

## 4b

In [ ]:
-- Profiling Results:

--  TBD

## 4c

## 4d